go-controller/pkg/node/gateway_init.go

```diff
@@ -82,15 +82,79 @@ func getIPv4Address(iface string) (*net.IPNet, error) {
 	if err != nil {
 		return nil, err
 	}
+	var ips []*net.IPNet
+	var foundIPv4 bool
+	var foundIPv6 bool
 	for _, addr := range addrs {
 		switch ip := addr.(type) {
 		case *net.IPNet:
-			if !utilnet.IsIPv6CIDR(ip) {
-				return ip, nil
+			if utilnet.IsIPv6CIDR(ip) {
+				if config.IPv6Mode && !foundIPv6 {
+					ips = append(ips, ip)
+					foundIPv6 = true
+				}
+			} else if config.IPv4Mode && !foundIPv4 {
+				ips = append(ips, ip)
+				foundIPv4 = true
 			}
 		}
 	}
-	return nil, nil
+	if config.IPv4Mode && !foundIPv4 {
+		return nil, fmt.Errorf("failed to find IPv4 address on interface %s", iface)
+	} else if config.IPv6Mode && !foundIPv6 {
+		return nil, fmt.Errorf("failed to find IPv6 address on interface %s", iface)
+	}
+	return ips, nil
+}
+
+func getGatewayNextHops() ([]net.IP, string, error) {
+	var gatewayNextHops []net.IP
+	var needIPv4NextHop bool
+	var needIPv6NextHop bool
+
+	if config.IPv4Mode {
+		needIPv4NextHop = true
+	}
+	if config.IPv6Mode {
+		needIPv6NextHop = true
+	}
+
+	// FIXME DUAL-STACK: config.Gateway.NextHop should be a slice of nexthops
+	if config.Gateway.NextHop != "" {
+		if config.IPv4Mode && !utilnet.IsIPv6String(config.Gateway.NextHop) {
+			gatewayNextHops = append(gatewayNextHops, net.ParseIP(config.Gateway.NextHop))
```

Comment:
should probably parse `config.Gateway.NextHop` first (after checking it's not `""`), and return an error if it doesn't parse. (The current code doesn't do this; it ends up treating an invalid NextHop the same as not specifying NextHop.)